frontend/src/components/pages/RegistrantExperience/PersonalInfo/personalInfoReducer.tsx

```diff
@@ -0,0 +1,166 @@
+import React from "react";
+import {
+  AddCamper,
+  DeleteCamper,
+  PersonalInfoActions,
+  PersonalInfoReducerDispatch,
+  UpdateCamper,
+  UpdateContact,
+} from "../../../../types/PersonalInfoTypes";
+import { RegistrantExperienceCamper } from "../../../../types/CamperTypes";
+
+export const CamperReducer = (
+  setCampers: React.Dispatch<
+    React.SetStateAction<RegistrantExperienceCamper[]>
+  >,
+  action: PersonalInfoReducerDispatch,
+): void => {
+  setCampers((campers: RegistrantExperienceCamper[]) => {
+    const newCampers: RegistrantExperienceCamper[] = JSON.parse(
+      JSON.stringify(campers),
+    ); // Deep Copy
+    switch (action.type) {
+      case PersonalInfoActions.ADD_CAMPER: {
+        const { campSessions } = action as AddCamper;
+        // Add a unique camper entry for all registered camp sessions
+        /* eslint-disable-next-line */
+        const campSessionIds = Array.from(campSessions.map((campSession) => campSession.id));
+        newCampers.push({
+          id: "",
+          campSessions: campSessionIds,
+          firstName: "",
+          lastName: "",
+          age: NaN,
+          contacts: [],
+          registrationDate: new Date(),
+          hasPaid: false,
+          chargeId: "",
+          charges: {
+            camp: NaN,
+            earlyDropoff: NaN,
+            latePickup: NaN,
+          },
+          optionalClauses: [],
+        });
+
+        // inject contact info
+        /* eslint-disable-next-line */
+            for (const contact of newCampers[0].contacts){
+          newCampers[newCampers.length - 1].contacts.push(
+            JSON.parse(JSON.stringify(contact)),
+          ); // Deep copy the contact
+        }
+        break;
+      }
+      case PersonalInfoActions.DELETE_CAMPER: {
+        const { camperId } = action as DeleteCamper;
+        newCampers.splice(camperId, 1);
+        break;
+      }
+      case PersonalInfoActions.UPDATE_CAMPER: {
+        const { camperId, field, data } = action as UpdateCamper;
+        if (field === "firstName") newCampers[camperId][field] = data as string;
+        else if (field === "lastName")
+          newCampers[camperId][field] = data as string;
+        else if (field === "age") newCampers[camperId][field] = data as number;
+        else if (field === "allergies")
+          newCampers[camperId][field] = data as string;
+        else if (field === "specialNeeds")
+          newCampers[camperId][field] = data as string;
+        break;
+      }
+      case PersonalInfoActions.UPDATE_CONTACT: {
+        const { contactId, field, data } = action as UpdateContact;
+        /* eslint-disable-next-line */
+        for (const camper of newCampers) {
```

Comment:
forEach here?

Reply:
The issue I have with using forEach is that I get this error

<img width="1148" alt="Screen Shot 2022-12-19 at 3 56 45 PM" src="https://user-images.githubusercontent.com/117593511/208522497-95fecc12-659b-4b5c-ba41-8f788bca48aa.png">